README.md
TabCAT
======

TabCAT (Tablet-based Cognitive Assessment Tool) is used by the [UCSF Memory and Aging Center](http://mac.ucsf.edu) to help diagnose Alzheimer's and other forms of dementia with a series of cognitive tasks, administered on an iPad or other tablet.

TabCAT is pure HTML5, so it can, in theory, run in any modern browser, and writing new tasks requires only basic web programming skills. It uses [CouchDB](http://couchdb.apache.org/) as both its web server and its backend to store transcripts of tasks, which can be retrieved in JSON format.

TabCAT is designed with [HIPAA](http://www.hhs.gov/ocr/privacy/) compliance in mind. By default, it does not store any [PHI](http://www.hhs.gov/ocr/privacy/hipaa/understanding/coveredentities/De-identification/guidance.html#protected) at all, though it can be configured to store limited dataset PHI (e.g. dates) or full PHI (see **Configuring TabCAT**, below.) Any PHI stored is always clearly tagged as such so that it can be later stripped out if need be.

TabCAT can deal with a poor or nonexistent network connection by automatically shifting into offline mode; data for tasks will be stored in the browser's local storage until the network is available again.

Installation
------------

### Development Environment

A packaged development environment is now provided using VirtualBox and Vagrant.  The instructions here do not replace the instructions for installation of the server, but merely provide supplement:

 * Download and install [Vagrant](https://www.vagrantup.com/downloads.html)
 * Download and install [VirtualBox](https://www.virtualbox.org/wiki/Downloads)
 * Clone the source into a directory of your choice.  A .gitattributes file has been added to ensure proper line endings, but you may need to additionally perform the following command on the host machine for the build system to work correctly:
```sh
git config --global core.autocrlf true
```
 * Open a console and navigate to the directory of the cloned source. If on Windows, this console may need to be run as an administrator in order for symlinks to not break the build.  Additional Virtualbox settings may be required if the build breaks due to symlink issues.
 * Run the following command:  
```sh
vagrant up
```
 * Optionally, from this directory, a private key is provided for accessing the virtual machine.  The location of it is ./.vagrant/machines/default/virtualbox/private_key.  If using PuTTY, you will need to create a .ppk from the private_key by using PuTTYgen before it can be used to log in.
 * Using keypair authentication, log into the virtual machine using an SSH client of choice.  If ssh is found as a binary in your path, you can simply perform "vagrant ssh" and it will attempt to log you in. By default, the port for SSH on the VM is 2222, although Vagrant will attempt to re-assign it if it is already in use.  The user to authenticate is "vagrant", and if not using keypair authentication, the password is also "vagrant".  This user has sudo privileges with no password required.
 * To start the couchdb server, in an instance of an SSH session, perform the following command.  This will kick off CouchDB as a background process.
```sh
sudo couchdb -b
``` 

### Dependencies

On the server side, TabCAT requires CouchDB ([installation instructions](http://docs.couchdb.org/en/latest/install/index.html)). We have been running on version 1.3.0, but anything later than 1.2.0 should work.

If you don't want to host a server, [Iris Couch](http://www.iriscouch.com/) is a good, free option.

In order to build and install TabCAT, you will need:

 * [node.js](http://nodejs.org/)
 * [kanso](http://kan.so)
 * [CoffeeScript](http://coffeescript.org/)
 * [UglifyJS](https://github.com/mishoo/UglifyJS2)
 * [curl](http://curl.haxx.se/download.html)
 * [GNU make](http://www.gnu.org/software/make/)

Don't panic! Once you have node.js, installing kanso, coffeescript, and UglifyJS is just a matter of running:

```sh
sudo npm install -g kanso coffee-script uglify-js
```

If you are on a UNIX system, you probably have `make` (and `curl`) already.

On Mac OS X, `make` is available as part of [XCode](https://developer.apple.com/xcode/); once you have XCode installed, install Command Line Tools from the Components tab of the Downloads preferences panel.

Due to issues with symbolic links in the build process, building on Windows is currently unsupported.

Finally, if you want to get involved in developing TabCAT, we recommend installing coffeelint (`sudo npm install -g coffeelint`).

### Setting up your server

The `setup_couchdb.sh` script will perform most of the setup needed for Couch DB.

```sh
export COUCHDB_URL=<COUCHDB-URL>
./setup_couchdb.sh
```
Under the hood, this script performs the following operations.

* Under the `couch_httpd_auth` section, sets `allow_persistent_cookies` to `true` and `timeout` to `3600`.
* Under the `uuids` section, sets `algorithm` to `random`.
* Creates an admin user whose username is read from the user.
* Creates a `tabcat` user with a password read from the user. This password is also stored in the `.tabcat_password` file.
* Creates a database named `tabcat`. Adds the `tabcat` user to the admins list. Adds a user provided email to members list.
* Creates a database named `tabcat-data`. Adds the `tabcat` user to the admins list. Adds a user provided email to members list.
* Under the `httpd` section, sets `bind_address` to `0.0.0.0`, so as to allow access to your server from other hosts.

You can also perform these operations from [Futon](http://wiki.apache.org/couchdb/Getting_started_with_Futon).

It's probably a good idea to force users to access your server via SSL; see [How to Enable SSL](http://wiki.apache.org/couchdb/How_to_enable_SSL) for further instructions.

### Adding additional users

New users will have to click the "Sign Up" link in Futon, using their (lowercase) email address as their username. Once they've signed up, enable their account by going to the `tabcat-data` database, clicking on Security, and adding their email to the Names section of members, just like you did with your own email: `["new.user@somewhere.com", "your.email@somewhere.com"]`.

### Installing TabCAT

`cd` to the root of the source directory.

Set the environment variable `TABCAT_HOST` to your web server's URL, with the `tabcat` user's credentials. That will look something like this:

`export TABCAT_HOST=http://tabcat:your-password@localhost:5984`

Then run `make`.

If you're developing TabCAT and you get symlink issues inside the Vagrant machine, run the command `vagrant-make` in the TabCAT source folder instead to build the project from the Vagrant box's home directory.

Additionally, if you're developing TabCAT, you may want to disable minification/uglification of assets by running the following command before `make`:

`export TABCAT_DEBUG=1`

This will allow you to see the uncompressed versions of css and javascript in the web browser.

### Configuring TabCAT

By default, TabCAT will not store any PHI at all, including dates
(other than year).

To change this, add a config file `/tabcat-data/config`. The easiest way to create this is through CouchDB's Futon interface; go to `/_utils/document.html?tabcat-data` and set the new document's ID to "config". To turn on dates, add a field `limitedPHI`, and set it to `true`. This enables storing of dates.

There is also a field `PHI`, which if set to true, allows storing of any PHI (e.g. patient name, voice recordings). Currently, TabCAT doesn't make use of this feature, though it may in the future.

### Using TabCAT

To access TabCAT, browse to `/tabcat/_design/console/login.html` on your server.

If you're on a iPad, click "Add to Home Screen" under the bookmarks menu (upper left), hit the home button, and re-launch the browser from the shortcut you just created. This launches TabCAT in Standalone mode (no nav bar), which is what you want for running tasks.

Then log in with your email and password. It'll prompt you for a patient code; if you're just testing things out, we recommend you use the patient code `0`.

If you're on Android, the best way to get the same kind of "standalone" behavior is to install [Firefox for Android](https://play.google.com/store/apps/details?id=org.mozilla.firefox). Add TabCAT to your home screen by clicking on the nav bar, the History tab, holding down on the entry for TabCAT (which should be at the top), and then choosing Add to Home Screen. Then, from Firefox, install [Full Screen mobile extension](https://addons.mozilla.org/En-us/mobile/addon/full-screen-252573/) and enable it from the Menu button.

If you're on a Microsoft Surface, your best bet is to launch TabCAT from the desktop in Internet Explore, and then enable Full Screen mode (F11, I think?).


CHANGES.txt
v0.9.1, 2015-11-12 -- Remaining Spanish Translations
- core:
  - Added more Spanish translations
- tasks:
  - digit-symbol:
    - Added Spanish instructions and navigation
  - memory:
    - Added back button to recall screens
    - Added in Latin version of pairings

v0.9.0, 2015-10-23 -- Alternate forms for Memory and Digit Symbol, Spanish translations
- build system:
  - Normalized line endings
  - Added vagrant-make to bypass symlink issues on shared folders
  - Began conversion from make to gulp
  - Fixed initial permissions on CouchDB setup
  - Added SASS support
- core:
  - Began style guide implementation
  - Added more internationalization support for Spanish translation
- console:
  - Added support for alternate forms for tasks
- reports:
  - Fixed bugs in line tasks reports
- tasks:
  - digit-symbol:
    - Added alternate forms
  - memory:
    - Added alternate forms
    - Added scoring capture
    - Added fade between faces
    - Changed task instructions

v0.8.9, 2015-08-27 -- Added Digit Symbol reporting and fix touch-related bugs
- core:
  - Extended jQuery to accommodate customized touch handling
- reports:
  - Added summary report for digit symbol
- tasks:
  - digit-symbol:
    - Added enhanced touch support
  - memory:
    - Minor visual enhancements
    - Added enhanced touch support

v0.8.8, 2015-08-01 -- Added Digit Symbol, Learning, and Delay tasks
- tasks:
  - line-perception:
    - updated norms
    - fixed catch trials
  - digit-symbol:
    - added Digit Symbol task
  - memory:
    - added Learning and Delay tasks
  - flanker:
    - fixed data export

v0.8.7, 2015-06-26 -- Added catch trials for on-task measure
- tasks:
  - line-perception:
    - added catch trials for on-task measures
- reports:
  - line perception reports now include catch trial data

v0.8.6, 2015-06-24 -- Added refined scoring
- tasks:
  - line-perception:
    - added refined scoring for high performers

v0.8.5, 2015-06-22 -- Update Line Perception Norms
- build system:
  - fixed bug in initial couchdb setup script
  - added initial dev environment setup with vagrant
- tasks:
  - line-perception:
    - updated norms

v0.8.4, 2015-02-08 -- EXAMINER upgrade
- build system:
  - single shared kanso packages directory (#129)
  - make can handle '%' in $TABCAT_HOST
- core:
  - TabCAT.UI.linkEmToPercentOfHeight() can now link to another element
  - fix validate_doc_update when secObjs.admins is undefined (#164)
- console:
  - higher login page layout, so keyboard doesn't cover fields (#163)
  - track aborted tasks, show in task picker (#167)
- tasks:
  - flanker:
    - color changes (#175)
    - better instructions (#174)
      - i18n temporarily disabled (need new translations)
    - always run test, even if flunked practice (#173)
    - hide fixation point (+) when showing arrows (#171)
  - all examiner tasks: convert PNGs to SVGs (#115)

v0.8.3, 2014-12-02 -- Timeout for login
- core:
  - login/logout functions can now time out (#159)
  - store timezoneOffset in encounter
- console:
  - UI feedback when login/logout button pushed (#159)
  - warn that logging out bypasses administration notes (#139)
- documentation:
  - explained how to enable (limited) PHI in README

v0.8.2, 2014-11-19 -- Replication
- build system:
  - single js/task.js file for tasks if you included it in kanso.json (#143)
- core:
  - security model re-enabled (#149)
  - notDesignDoc filter for replicating data only (#145)
  - only admins can change configs
- console:
  - TabCAT logo takes up entire vertical space in status bar (#150)
- setup:
  - setup_couchdb.sh now configures everything execpt SSL (#25)
- tasks:
  - line-perception and dart now use js/task.js

v0.8.1, 2014-10-15 -- Fixes for iPad 1
- core:
  - added ECMAScript 5 shims so that gauss lib works on iPad 1 (#146)
  - if task scoring fails, don't hang, and mark task as completed

v0.8.0, 2014-10-08 -- Task scoring
- core:
  - scoring (#121)
    - CouchDB backend can score eventLogs from tasks
    - tasks also score themselves into localStorage for offline mode
      - encounterTaskScoring (localStorage) replaces encounterTasksFinished
  - security model no longer applies to admins, for replication (#130)
  - make pushes to all tabcat-data* DBs on server (#123)
  - i18n (#6)
    - (there is not yet a way to change language in TabCAT interface)
    - translations for some tasks -- some happened in 0.7.0
    - translations for task names
- console:
  - use professional TabCAT logo (#133)
    - background is now "cat navy blue"
  - partially completed, hidden patient-scoring page
- reports:
  - line-tasks report has correct filename (#117)
- tasks
  - line-perception:
    - scoring (#121)
    - chinese translations (#5)
  - stargazer: change to color of font for "ONE" (#119)

v0.7.0, 2014-06-13 -- More EXAMINER, Console Split
- core:
  - added gauss library to vendor/ (currently not available in tasks)
  - console split out (see below)
- console:
  - console has its own design document (console/) (#68)
    - login page is /now tabcat/_design/console/login.html
    - some otherwise unused TabCAT library methods are now only in console
- reports:
  - added date and version columns to line-tasks report (#112)
  - added data quality fields to line-tasks and stargazer reports (#113)
  - factored some common code into core/js/adhoc/report.coffee
  - flanker now has summary and detailed report
  - new cpt and set-shifting reports (summary and detailed)
- tasks:
  - flanker:
    - small updates to label formatting
    - added two throwaway trials between practice and testing
  - cpt: new task!
  - set-shifting: new task!

v0.6.3, 2014-05-02 -- Flanker
- core library:
  - Task.logEvent() uses $.now() when event.timeStamp is unreliable (#102)
- tasks:
  - flanker: new task! includes translations
  - stargazer: correct version (v0.6.2)

v0.6.2, 2014-03-27 -- Revert stargazer staircasing
- build system:
  - kanso packages directories are included, since kan.so is down (#96)

setup_couchdb.sh
#!/usr/bin/env bash

# Bash script to set up Couch DB for Tabcat
# Expects COUCHDB_URL environment variable to be set to point to Couch DB URL. E.g. http://127.0.0.1:5984

CONFIG_URL=$COUCHDB_URL/_config
HTTPD_AUTH_SECTION=$CONFIG_URL/couch_httpd_auth
UUIDS_SECTION=$CONFIG_URL/uuids
ADMINS_SECTION=$CONFIG_URL/admins
USERS_SECTION=$COUCHDB_URL/_users
HTTPD_SECTION=$CONFIG_URL/httpd

echo "Creating admin user ..."
echo "Please enter Admin user's id:"
read ADMIN_USER
echo "Please enter Admin user's password:"
read -s ADMIN_PASSWORD
curl -X PUT $ADMINS_SECTION/$ADMIN_USER -d '"'$ADMIN_PASSWORD'"'

AUTH_STRING=$ADMIN_USER:$ADMIN_PASSWORD

echo "Creating tabcat user ..."
TABCAT_PASSWORD=$(cat /dev/urandom | env LC_CTYPE=C tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1)
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $USERS_SECTION/org.couchdb.user:tabcat -d '{"name":"tabcat", "password": "'$TABCAT_PASSWORD'", "type": "user", "roles": []}'
echo $TABCAT_PASSWORD > .tabcat_password

#adding sandbox user by default
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $USERS_SECTION/org.couchdb.user:s@ndbox -d '{"name":"s@ndbox", "password": "s@ndbox", "type": "user", "roles": []}'

echo "Please enter database user's email address:"
read USER_EMAIL

for db in tabcat tabcat-data
do
    echo "Creating '$db' database with correct permissions..."
    curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $COUCHDB_URL/$db
done

#the tabcat database needs an empty members section
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $COUCHDB_URL/tabcat/_security -d '{"admins":{"names":["tabcat"],"roles":["admins"]}}'
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $COUCHDB_URL/tabcat-data/_security -d '{"admins":{"names":["tabcat"],"roles":["admins"]},"members":{"names":["'$USER_EMAIL'", "s@ndbox"]}}'

echo "Configuring couch_httpd_auth ..."
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $HTTPD_AUTH_SECTION/allow_persistent_cookies -d '"true"'
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $HTTPD_AUTH_SECTION/timeout -d '"3600"'

echo "Configuring httpd ..."
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $HTTPD_SECTION/bind_address -d '"0.0.0.0"'

echo "Configuring uuids ..."
curl -u $AUTH_STRING --header "Content-Type: application/json" -X PUT $UUIDS_SECTION/algorithm -d '"random"'
